chalk-solve/src/clauses.rs

```diff
@@ -121,21 +121,23 @@ pub(crate) fn program_clauses_for_goal<'db, I: Interner>(
     );
     let interner = db.interner();
 
-    // FIXME: change this to use `.chain().filter()`
-    let mut vec = vec![];
-    vec.extend(db.custom_clauses());
-    program_clauses_that_could_match(db, environment, goal, &mut vec)?;
-    vec.retain(|c| c.could_match(interner, goal));
-    vec.extend(
-        db.program_clauses_for_env(environment)
-            .iter(interner)
-            .filter(|c| (*c).could_match(interner, goal))
-            .cloned(),
-    );
+    let custom_clauses = db.custom_clauses().into_iter();
+    let clauses_that_could_match =
+        program_clauses_that_could_match(db, environment, goal).map(|cl| cl.into_iter())?;
+
+    let clauses: Vec<ProgramClause<I>> = custom_clauses
+        .chain(clauses_that_could_match)
+        .chain(
+            db.program_clauses_for_env(environment)
+                .iter(interner)
+                .cloned(),
```

Comment:
We can avoid the cloning here if we don't abstract away the `Vec` via `InternedProgramClauses`. Why do we do this?

Reply:
Hm, not sure really. @jackh726 Any idea why this is the case ?

Reply:
Why do we do what exactly? Return `InternedProgramClauses` from `program_clauses_for_env`? Or just use `InternedProgramClauses` instead of `Vec`? Either way, we don't want `Vec`s, we want the implementor of `Interner` to decide what we use (maybe it's `SmallVec`s, maybe it's an interned slice). Cloning here is fine, since we also say that cloning a `ProgramClause` is fine.

Reply:
Got it, thanks !

Reply:
Hm, maybe we could add an `IntoIterator<Item = InternedProgramClause>` bound on the type?